README.md
# EasyPost Node Client Library

EasyPost is a simple shipping API. You can sign up for an account at https://easypost.com.

[![CI](https://github.com/EasyPost/easypost-node/workflows/CI/badge.svg)](https://github.com/EasyPost/easypost-node/actions?query=workflow%3ACI)
[![npm version](https://badge.fury.io/js/%40easypost%2Fapi.svg)](https://badge.fury.io/js/%40easypost%2Fapi)

## Installation

```bash
npm install --save @easypost/api
```

**Note:** if you are using a version of Node less than 6.9, you will need to install and
include a polyfill, such as `babel-polyfill`, and include it in your project:

```bash
npm install --save babel-polyfill
```

In your file:

```javascript
require('babel-polyfill');
const EasyPost = require('@easypost/api');
```

## Compatability

By default, @easypost/api works with Node v6 LTS. To include for other versions
of node, you can use:

* `require('@easypost/api/easypost.8-lts.js')` (Node 8.9+)
* `require('@easypost/api/easypost.6-lts.js')` (Node 6.9+)
* `require('@easypost/api/easypost.legacy.js')` (Node 0.10+)

## Example

```javascript
const apiKey = 'cueqNZUb3ldeWTNX7MU3Mel8UXtaAMUi';
const EasyPost = require('@easypost/api');

const api = new EasyPost(apiKey);

// set addresses
const toAddress = new api.Address({
  name: 'Dr. Steve Brule',
  street1: '179 N Harbor Dr',
  city: 'Redondo Beach',
  state: 'CA',
  zip: '90277',
  country: 'US',
  phone: '310-808-5243'
});

const fromAddress = new api.Address({
  name: 'EasyPost',
  street1: '118 2nd Street',
  street2: '4th Floor',
  city: 'San Francisco',
  state: 'CA',
  zip: '94105',
  phone: '415-123-4567'
});

/* es5 with promises: */
fromAddress.save().then(addr => {
  console.log(addr.id);
});

/* es2017 with async/await: */
await fromAddress.save();
console.log(fromAddress.id);
```

### Options

You can construct an API instance with certain options:

```javascript
const api = new Api("mykey", {
  timeout: 120000,
  baseUrl: "https://api.easypost.com/v2/",
  useProxy: false,
  superagentMiddleware: s => s,
  requestMiddleware: r => r,
});
```

#### timeout

Time in milliseconds that should fail requests.

#### baseUrl

Change the base URL that the API library uses. Useful if you proxy requests
from a frontend through a server.

#### useProxy

Disable using the API key. Useful if you proxy requests from a frontend through
a server.

#### superagentMiddleware

Function that takes `superagent` and returns `superagent`. Useful if you need
to wrap superagent in a function, such as many superagent libraries do.

```javascript
import superagentLib from 'some-superagent-lib';

const api = new Api("my-key", {
  superagentMiddleware: s => superagentLib(s),
});
```

#### requestMiddleware

Function that takes a superagent `request` and returns that request. Useful if
you need to hook into a request:

```javascript
import superagentLib from 'some-superagent-lib';

const api = new Api("my-key", {
  requestMiddleware: r => {
    r.someLibFunction(SOME_CONFIG_VALUE);
    return r;
  },
});
```

## Development

To test out the API, you can run the following:

```bash
npm install -g @easypost/api
easypost
```

### Install Dependencies

```bash
npm install
```

### Build

Build the various versions of the client library by running the following:

```bash
npm run build
```

### Testing

Run unit tests by running the following:

```bash
npm test
```

### Interactive CLI

Replace `easypost.js` with whatever compatabile version you wish, as defined
under `Compatibility`.

```bash
API_KEY=yourkey ./repl.js --local easypost.js
```

## Releasing

1. Update the version in the `package.json` file
1. Update the `CHANGELOG` file
1. Ensure that checked-in compiled assets are re-built with `npm run build`
1. Tag the release on GitHub
1. Publish the updated npm package

## Note on ES6 Usage

You can import specific versions of the compiled code if you're using later
versions of Node. 

```javascript
// Imports the un-transformed es6
import "@easypost/api/src/easypost"

// Use the following to import mininally transformed versions
import "@easypost/api/easypost.6-lts"
import "@easypost/api/easypost.8-lts"
import "@easypost/api/easypost.legacy.js" // (v0.10)
```


CHANGELOG
=== NEXT
- Restored support for Node 6.x
- Bumped many dependencies
- Swapped out Travis CI for GitHub Actions

=== 3.9.1 2020-06-09
- Bump `websocket-extensions` to 0.1.4

=== 3.9.0 2020-05-15

* Added the ability to retrieve individual user's API keys
* Added the ability to retrieve all address records
* Removed the unusable shipment return method
* Security updates to Node modules
* Swapped link for Travis CI from .org to .com
* Updated stale unit tests

=== 2.1.1 2016-07-29

* Added Insurance object support


=== 2.1.0 2016-06-03

* Added an optional timeout for requests. Defaults to 120000ms (thanks Shyp and misterMuyiwa!)


=== 2.0.10 2016-03-24

* Added Orders
* Fixed some bad tests (Addresses, Shipments, Parcels)


=== 2.0.9 2016-02-03

* Added "verify" and "verify_strict" options to Address create, and made tests for them.


=== 2.0.8 2015-11-05

* Updated README and tests


=== 2.0.7 2015-07-14

* Changed shipment rate request to use GET not POST
* Fixed some tests


=== 2.0.6 2015-02-13

* Added ability to buy and cancel pickups


=== 2.0.5 2014-11-04

* Added tracker to shipment buy response
* Added User-Agent fields to request header
* Updated tests


=== 2.0.4 2013-12-29

* Moved vows module to a devDependancy.


=== 2.0.3 2013-12-29

* Fixed some bugs and improved Shipment.lowestRate by allowing negative filters.
* Improved handling of the partial shipment response object that Shipment.buy returns.
* Minor syntax and documentation improvements.


=== 2.0.2 2013-10-13

* Event resource added for webhook consumption.
* Tracker resource added.
* Batch.create_scan_form added to request a ScanForm containing all of a Batch's shipments.


=== 2.0.1 2013-07-05

* Added label function to Shipment to request specific label file_formats (pdf, epl2, zpl).
* Added insure function to Shipment. Add insurance to any shipment in one call!
* Added create_and_verify function to Address.


repl.js
#!/usr/bin/env node

/* eslint import/no-extraneous-dependencies: 0, global-require: 0, import/no-unresolved: 0, no-console: 0, max-len: 0 */

require('source-map-support').install();
require('core-js/stable');

const args = require('yargs').argv;

const repl = require('repl');

process.on('unhandledRejection', (err) => {
  console.log(err, err.stack);
});

let packageInfo;
let API;

if (args.local) {
  packageInfo = require('./package.json');
  API = require(`./${args.local}`).default;
} else {
  packageInfo = require('@easypost/api/package.json');
  API = require('@easypost/api');
}

console.log(`Starting ${packageInfo.name} v${packageInfo.version} repl`);
console.log('Enter `help()` for information.');


let api;

if (process.env.API_KEY) {
  api = new API(process.env.API_KEY);
} else {
  console.log([
    'Create an instance by using `api = new API(apikey)`, or restart',
    'the repl with an API_KEY environment variable.',
  ].join(' '));
}

const local = repl.start('$> ');

function help() {
  const helpText = [
    'To try out the API, use the available instance of `api` to make requests.',
    'For example, try writing: `api.Address.all();`',
    'Sample data is also available: toAddress, fromAddress.',
  ].join(' ');

  console.log(helpText);
}

local.context.toAddress = {
  name: 'Dr. Steve Brule',
  street1: '179 N Harbor Dr',
  city: 'Redondo Beach',
  state: 'CA',
  zip: '90277',
  country: 'US',
  phone: '310-808-5243',
};

local.context.fromAddress = {
  name: 'EasyPost',
  street1: '118 2nd Street',
  street2: '4th Floor',
  city: 'San Francisco',
  state: 'CA',
  zip: '94105',
  phone: '415-123-4567',
};

local.context.badAddress = {
  verify: ['delivery'],
  street1: 'UNDELIVERABLE ST',
  city: 'SAN FRANCISCO',
  state: 'CA',
  zip: '94104',
  country: 'US',
  company: 'EasyPost',
  phone: '415-123-4567',
};

local.context.API = API;
local.context.api = api;
local.context.help = help;
